pkg/download/buffer.go

```diff
@@ -64,42 +57,37 @@ type firstReqResult struct {
 func (m *BufferMode) Fetch(ctx context.Context, url string) (io.Reader, int64, error) {
 	logger := logging.GetLogger()
 
-	br := newBufferedReader(m.pool)
+	firstChunk := newBufferedReader(m.pool)
 
 	firstReqResultCh := make(chan firstReqResult)
-	m.queue.submit(func() {
-		m.sem.Go(func() error {
-			defer close(firstReqResultCh)
-			defer br.Done()
-			firstChunkResp, err := m.DoRequest(ctx, 0, m.chunkSize()-1, url)
-			if err != nil {
-				firstReqResultCh <- firstReqResult{err: err}
-				return err
-			}
-
-			defer firstChunkResp.Body.Close()
-
-			trueURL := firstChunkResp.Request.URL.String()
-			if trueURL != url {
-				logger.Info().Str("url", url).Str("redirect_url", trueURL).Msg("Redirect")
-			}
-
-			fileSize, err := m.getFileSizeFromContentRange(firstChunkResp.Header.Get("Content-Range"))
-			if err != nil {
-				firstReqResultCh <- firstReqResult{err: err}
-				return err
-			}
-			firstReqResultCh <- firstReqResult{fileSize: fileSize, trueURL: trueURL}
-
-			contentLength := firstChunkResp.ContentLength
-			n, err := br.ReadFrom(firstChunkResp.Body)
-			if err != nil {
-				return err
-			} else if n != contentLength {
-				return ErrContentLengthMismatch{downloadedBytes: n, contentLength: contentLength}
-			}
-			return nil
-		})
+	m.queue.submitLow(func() {
+		defer close(firstReqResultCh)
```

Comment:
I think I probably need to sit down with you to really grok what this code is doing, but there seems to be a pretty significant change between the old code and the new code, which is that the enqueued "work" items here actually executed in parallel before (because `m.sem.Go` returns immediately), whereas now they execute serially. Doesn't that matter?

Reply:
let's discuss.